### ISSUE TITLE
Foundry out of box support

### ISSUE THREAD

**Issue by chitalian** [juanfranblanco/vscode-solidity#335]
Steps to reproduce
```bash
forge init solidity-tutorial
code .
```

in the test file
![image](https://user-images.githubusercontent.com/26822232/162125105-b797089d-a1ec-4d1f-a1ca-b8e4757b27d8.png)

**Comment by dcposch** [comment]
I ran into the same problem today.

### Workarounds

I found the following both work:
- Open the Foundry project (eg `packages/contracts`) in VS Code, rather than the monorepo.
- Add `"solidity.packageDefaultDependenciesDirectory": "packages/portal-contracts/lib"`, `"solidity.packageDefaultDependenciesContractsDirectory": "src"`.

Transitive dependencies are still broken. So `forge-std` works since it's in `lib/forge-std`, but `ds-test` does not work, since it's in `lib/forge-std/lib/ds-test`.

### Real fix

I think we could solve for monorepos (foundry or otherwise) in one of a few ways.

One might be to add a "solidity root" setting. But this extension has a lot of knobs already, and the new user experience of trying to Google how to set them to make your project compile is not very pleasant.

An alternative would be to go in the direction of removing knobs. Is there a dependency resolution strategy that would reduce the need for `packageDefaultDependencies...` and `remappings.txt`?

If there's rough consensus on how to do this, I'm happy to contribute a PR.

**Comment by 0xdeployer** [comment]
Hi anyone have updates on this?

**Comment by holic** [comment]
Just want to add a bit more context/another requirement: a monorepo may have more than one contracts directory/package (foundry project), and would be good to be able to resolve the package-local remappings.txt (rather than a global singleton).

**Comment by holic** [comment]
For a little more context, here's my workaround for a pnpm monorepo that uses only pnpm-managed dependencies (no foundry deps):
https://github.com/latticexyz/mud/commit/f238f67834c2062fdaf672003e6466b6e33488a0

Confusingly, one dependency we use that isn't specified in these remappings still works fine out-of-the-box. Maybe because it's just used by one package and not multiple, so there's no remapping conflicts?
https://github.com/latticexyz/mud/blob/main/packages/solecs/src/Ownable.sol#L4-L5

**Comment by juanfranblanco** [comment]
@holic thank you, I'll have a look !

**Comment by holic** [comment]
I believe "nearest remapping" will do the trick, yep! That should help with both the pnpm use case and with foundry libs, and still "bubble up" to the project root's remapping for compatibility with existing projects.

Ideally both `remappings.txt` and `foundry.toml` would be supported (unclear which would take precedence if both are found). I personally tend to use `remappings.txt` since I think it's more portable (not just a foundry thing)?

**Comment by juanfranblanco** [comment]
I think the idea of having a common project file is not going to happen anytime, so remappings it is and specific framework type.

**Comment by juanfranblanco** [comment]
![solidity-multiroot-support](https://user-images.githubusercontent.com/562371/231835580-6ea35095-5ede-4448-a02d-a55b9f1d8d9e.gif)
some updates

**Comment by holic** [comment]
Awesome! Just saw the commit to main.

A few questions:
- is this opt-in because it has to do a directory search and that can be slow to resolve?
- how often does the search happen?
- if a `foundry.toml` is encountered but that `foundry.toml` doesn't have `remappings`, does it fall back to `remappings.txt`? (we use both `foundry.toml` and `remappings.txt` in our projects)

**Comment by juanfranblanco** [comment]
@holic i have added foundry remappings too, let me know how it goes. This has to be enabled as it might add confusion. Let me know how it goes.. it is out on 151

**Comment by juanfranblanco** [comment]
@holic foundry takes preference over remappings if found.. now they could be merged instead and have both

**Comment by juanfranblanco** [comment]
Yes this is what is currently doing at the moment, so foundry if it has the remappings settings will take preference over the remappings.txt. And yes you are right better don't mess around with that, as it will create anti-patterns on projects. I guess best is to stick to one.

**Comment by holic** [comment]
Sounds great!

I just found this btw, which might help your file search: https://github.com/sindresorhus/find-up
```ts
const nearestFile = await findUp(projectFilesAtRoot);
```

**Comment by juanfranblanco** [comment]
Oh I crafted my own :)
